src/fsharp/TypeChecker.fs

```diff
@@ -7092,33 +7099,226 @@ and TcObjectExpr cenv overallTy env tpenv (synObjTy, argopt, binds, extraImpls,
 and TcConstStringExpr cenv overallTy env m tpenv s =
 
     if (AddCxTypeEqualsTypeUndoIfFailed env.DisplayEnv cenv.css m overallTy cenv.g.string_ty) then 
-      mkString cenv.g m s, tpenv
+        mkString cenv.g m s, tpenv
     else 
-      let aty = NewInferenceType ()
-      let bty = NewInferenceType ()
-      let cty = NewInferenceType ()
-      let dty = NewInferenceType ()
-      let ety = NewInferenceType ()
-      let ty' = mkPrintfFormatTy cenv.g aty bty cty dty ety
-      if (not (isObjTy cenv.g overallTy) && AddCxTypeMustSubsumeTypeUndoIfFailed env.DisplayEnv cenv.css m overallTy ty') then 
+        TcFormatStringExpr cenv overallTy env m tpenv s
+
+and TcFormatStringExpr cenv overallTy env m tpenv (fmtString: string) =
+    let g = cenv.g
+    let aty = NewInferenceType ()
+    let bty = NewInferenceType ()
+    let cty = NewInferenceType ()
+    let dty = NewInferenceType ()
+    let ety = NewInferenceType ()
+    let formatTy = mkPrintfFormatTy g aty bty cty dty ety
+
+    // This might qualify as a format string - check via a type directed rule
+    let ok = not (isObjTy g overallTy) && AddCxTypeMustSubsumeTypeUndoIfFailed env.DisplayEnv cenv.css m overallTy formatTy
+
+    if ok then
         // Parse the format string to work out the phantom types 
         let formatStringCheckContext = match cenv.tcSink.CurrentSink with None -> None | Some sink -> sink.FormatStringCheckContext
-        let normalizedString = (s.Replace("\r\n", "\n").Replace("\r", "\n"))
+        let normalizedString = (fmtString.Replace("\r\n", "\n").Replace("\r", "\n"))
```

Comment:
Is there a way to avoid this double string allocation?

Reply:
IIRC the checking of strings in general has a notable impact on overall memory usage and if we can be a little clever to avoid that here we should definitely take it. People will be using these things a lot.

Reply:
Would need string builder and a state machine like this:

```
i-2 i-1 i   action
# 3x
x   x   x   output x
# 1r
x   x   r   hold
x   r   x   output n x
r   x   x   output x
# 1n
x   x   n   output n
x   n   x   output x
n   x   x   output x
# 1r1n
x   r   n   output n
r   n   x   output x
n   x   r   hold
# 1n1r
x   n   r   hold
n   r   x   output n x
r   x   n   output n
# 2r
x   r   r   output n
r   x   r   hold
r   r   x   output n x
# 2n
x   n   n   output n
n   n   x   output x
n   x   n   output n
# 2r1n
r   r   n   output n
r   n   r   hold
n   r   r   output n
# 2n1r
n   n   r   hold
n   r   n   output n
r   n   n   output n
# 3n
n   n   n   output n
# 3r
r   r   r   output n
```

Reply:
> Is there a way to avoid this double string allocation?

In this case it's very low overhead compared to the surrounding allocations in general in all the string format checking code.